iroh-gossip/src/net/handles.rs

```diff
@@ -128,9 +130,22 @@ impl GossipReceiver {
     }
 
     /// Waits until we are connected to at least one node.
+    ///
+    /// This progresses the stream until we received [`GossipEvent::Joined`], which is the first
+    /// item emitted on the stream.
+    ///
+    /// Note that this consumes the [`GossipEvent::Joined`] event. If you want to act on these
+    /// initial neighbors, use [`Self::neighbors`] after awaiting [`Self::joined`].
     pub async fn joined(&mut self) -> Result<()> {
-        while self.neighbors.is_empty() {
-            let _ = self.try_next().await?;
+        if !self.joined {
+            match self
+                .try_next()
+                .await?
+                .context("Gossip receiver closed before Joined event was received.")?
+            {
+                Event::Gossip(GossipEvent::Joined(_)) => {}
```

Comment:
why are you throwing the initial neighbours here away?

Reply:
They are added to `self.neighbors` in `poll_next` already.